tests/python/json_ffi/test_json_ffi_engine_image.py

```diff
@@ -0,0 +1,382 @@
+# pylint: disable=chained-comparison,line-too-long,missing-docstring,
+# pylint: disable=too-many-arguments,too-many-locals,unused-argument,unused-variable
+import base64
+import json
+import queue
+import threading
+from typing import Any, Callable, Dict, Iterator, List, Literal, Optional, Union
+
+import requests
+import tvm
+
+from mlc_llm.protocol import openai_api_protocol
+from mlc_llm.serve import engine_utils
+from mlc_llm.serve.engine_base import (
+    EngineConfig,
+    SpeculativeMode,
+    _infer_kv_cache_config,
+    _parse_models,
+    _process_model_args,
+    detect_device,
+)
+from mlc_llm.tokenizer import Tokenizer
+
+
+def base64_encode_image(url: str) -> str:
+    response = requests.get(url)
+    response.raise_for_status()  # Ensure we got a successful response
+    image_data = base64.b64encode(response.content)
+    image_data_str = image_data.decode("utf-8")
+    data_url = f"data:image/jpeg;base64,{image_data_str}"
+    return data_url
+
+
+image_prompts = [
+    [
+        {
+            "role": "user",
+            "content": [
+                {
+                    "type": "image_url",
+                    "image_url": f"{base64_encode_image('https://llava-vl.github.io/static/images/view.jpg')}",
+                },
+                {"type": "text", "text": "What does the image represent?"},
+            ],
+        }
+    ]
+]
+
+
+def create_model_defined_generation_config(
+    temperature: float, top_p: float, frequency_penalty: float, presence_penalty: float
+) -> tvm.runtime.Object:
+    return tvm.get_global_func("mlc.json_ffi.ModelDefinedGenerationConfig")(
+        temperature,
+        top_p,
+        frequency_penalty,
+        presence_penalty,
+    )
+
+
+def create_json_ffi_engine_config(
+    conv_template: str, model_generation_cfgs: Dict[str, tvm.runtime.Object]
+) -> tvm.runtime.Object:
+    return tvm.get_global_func("mlc.json_ffi.JSONFFIEngineConfig")(
+        conv_template, model_generation_cfgs
+    )
+
+
+class EngineState:
+    sync_queue: queue.Queue
+
+    def get_request_stream_callback(self) -> Callable[[List[str]], None]:
+        # ChatCompletionStreamResponse
+
+        def _callback(chat_completion_stream_responses_json_str: List[str]) -> None:
+            self._sync_request_stream_callback(chat_completion_stream_responses_json_str)
+
+        return _callback
+
+    def _sync_request_stream_callback(
+        self, chat_completion_stream_responses_json_str: List[str]
+    ) -> None:
+        # Put the delta outputs to the queue in the unblocking way.
+        self.sync_queue.put_nowait(chat_completion_stream_responses_json_str)
+
+
+class JSONFFIEngine:
+    def __init__(  # pylint: disable=too-many-arguments,too-many-locals
```

Comment:
Let us consolidate the implementation of JSONFFIEngine  into a single place.

`mlc_llm/python/json_ffi/engine.py`

Then import these from there, so we don't have to duplicate things for json_ffi_engine_image and json_ffi testcases